core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -2190,6 +2209,85 @@ private Optional<OutputDataSizeEstimateResult> getEstimatedOutputDataSize()
             return Optional.of(new OutputDataSizeEstimateResult(new OutputDataSizeEstimate(estimateBuilder.build()), OutputDataSizeEstimateStatus.ESTIMATED_BY_PROGRESS));
         }
 
+        private Optional<OutputDataSizeEstimateResult> getEstimatedMarkDistinctStageOutputDataSize(Function<StageId, StageExecution> stageExecutionLookup)
+        {
+            if (!markDistinctChainEstimationEnabled) {
+                return Optional.empty();
+            }
+
+            if (!isSingleMarkDistinctStage(this.getStageInfo().getPlan())) {
+                return Optional.empty();
+            }
+
+            if (!isNoMorePartitions()) {
+                return Optional.empty();
+            }
+
+            PlanFragment planFragment = this.getStageInfo().getPlan();
+
+            List<RemoteSourceNode> remoteSourceNodes = planFragment.getRemoteSourceNodes();
+            long inputSizeEstimate = 0;
+            for (RemoteSourceNode remoteSourceNode : remoteSourceNodes) {
+                for (PlanFragmentId sourceFragmentId : remoteSourceNode.getSourceFragmentIds()) {
+                    StageId sourceStageId = StageId.create(queryStateMachine.getQueryId(), sourceFragmentId);
+
+                    StageExecution sourceStage = stageExecutionLookup.apply(sourceStageId);
+                    requireNonNull(sourceStage, "sourceStage is null");
```

Comment:
redundant, without explicit check same message would be used
(if we want a better message, we could cite sourceStageId in the message)